File: backend/src/modules/buildings/building.service.js
const prisma = require('../../lib/prisma');
const { HttpError } = require('../../lib/httpError');
const { sortBySortOrderThenName } = require('../../lib/hierarchySort');
const { normalizeEntityStatusForDb } = require('../siteHierarchy/siteHierarchy.service');

async function assertSiteExists(siteId) {
  const site = await prisma.site.findUnique({ where: { id: siteId } });
  if (!site) {
    throw new HttpError(404, 'Site not found');
  }
}

async function listBuildingsBySite(siteId) {
  await assertSiteExists(siteId);
  return sortBySortOrderThenName(
    await prisma.building.findMany({
      where: { siteId },
      orderBy: { name: 'asc' },
      include: { _count: { select: { floors: true } } },
    })
  );
}

async function getBuildingById(id) {
  const building = await prisma.building.findUnique({
    where: { id },
    include: {
      site: true,
      floors: { orderBy: { name: 'asc' } },
    },
  });
  if (!building) {
    throw new HttpError(404, 'Building not found');
  }
  if (building.floors) sortBySortOrderThenName(building.floors);
  return building;
}

async function createBuilding(siteId, data) {
  await assertSiteExists(siteId);
  const {
    name,
    addressLine1,
    addressLine2,
    city,
    state,
    postalCode,
    country,
    latitude,
    longitude,
  } = data;

  if (!name || !addressLine1 || !city || !state || !postalCode || !country) {
    throw new HttpError(
      400,
      'name, addressLine1, city, state, postalCode, and country are required'
    );
  }

  const sortOrder =
    data.sortOrder !== undefined && data.sortOrder !== null && data.sortOrder !== ''
      ? Number(data.sortOrder)
      : 0;
  return prisma.building.create({
    data: {
      siteId,
      name: String(name).trim(),
      addressLine1: String(addressLine1).trim(),
      addressLine2: addressLine2 != null ? String(addressLine2).trim() : null,
      city: String(city).trim(),
      state: String(state).trim(),
      postalCode: String(postalCode).trim(),
      country: String(country).trim(),
      latitude: latitude != null ? Number(latitude) : null,
      longitude: longitude != null ? Number(longitude) : null,
      ...(data.status !== undefined ? { status: normalizeEntityStatusForDb(data.status) } : {}),
      ...(data.buildingType != null && String(data.buildingType).trim()
        ? { buildingType: String(data.buildingType).trim() }
        : {}),
      ...(data.buildingCode != null && String(data.buildingCode).trim()
        ? { buildingCode: String(data.buildingCode).trim() }
        : {}),
      ...(data.description != null && String(data.description).trim()
        ? { description: String(data.description).trim() }
        : {}),
      sortOrder: Number.isFinite(sortOrder) ? sortOrder : 0,
    },
  });
}

async function updateBuilding(id, data) {
  await getBuildingById(id);
  const allowed = [
    'name',
    'addressLine1',
    'addressLine2',
    'city',
    'state',
    'postalCode',
    'country',
    'latitude',
    'longitude',
    'buildingType',
    'buildingCode',
    'description',
    'sortOrder',
  ];
  const update = {};
  for (const key of allowed) {
    if (data[key] !== undefined) {
      if (key === 'latitude' || key === 'longitude') {
        update[key] = data[key] == null ? null : Number(data[key]);
      } else if (key === 'sortOrder') {
        const n = Number(data[key]);
        update[key] = Number.isFinite(n) ? n : 0;
      } else if (key === 'addressLine2') {
        update[key] = data[key] == null ? null : String(data[key]).trim();
      } else if (key === 'buildingType' || key === 'buildingCode') {
        const s = data[key] == null ? '' : String(data[key]).trim();
        update[key] = s.length ? s : null;
      } else if (key === 'description') {
        const s = data[key] == null ? '' : String(data[key]).trim();
        update[key] = s.length ? s : null;
      } else {
        update[key] =
          typeof data[key] === 'string' ? data[key].trim() : data[key];
      }
    }
  }
  if (data.status !== undefined) {
    update.status = normalizeEntityStatusForDb(data.status);
  }
  if (Object.keys(update).length === 0) {
    throw new HttpError(400, 'No fields to update');
  }
  return prisma.building.update({
    where: { id },
    data: update,
  });
}

async function deleteBuilding(id) {
  await getBuildingById(id);
  return prisma.building.delete({
    where: { id },
  });
}

module.exports = {
  listBuildingsBySite,
  getBuildingById,
  createBuilding,
  updateBuilding,
  deleteBuilding,
};

File: server/src/main/java/com/legioncontrols/server/sites/BuildingService.java
package com.legioncontrols.server.sites;

import com.legioncontrols.server.common.ApiException;
import com.legioncontrols.server.jooq.generated.enums.Entitystatus;
import com.legioncontrols.server.jooq.generated.tables.records.BuildingRecord;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import org.jooq.DSLContext;
import org.jooq.Record2;
import org.springframework.stereotype.Service;

import static com.legioncontrols.server.jooq.generated.tables.Building.BUILDING;
import static com.legioncontrols.server.jooq.generated.tables.Floor.FLOOR;
import static com.legioncontrols.server.jooq.generated.tables.Site.SITE;
import static org.jooq.impl.DSL.count;

/** Java equivalent of backend/src/modules/buildings/building.service.js. */
@Service
public class BuildingService {

    private final DSLContext dsl;

    public BuildingService(DSLContext dsl) {
        this.dsl = dsl;
    }

    private void assertSiteExists(String siteId) {
        boolean exists = dsl.fetchExists(dsl.selectFrom(SITE).where(SITE.ID.eq(siteId)));
        if (!exists) {
            throw ApiException.notFound("Site not found");
        }
    }

    public List<BuildingDto> listBuildingsBySite(String siteId) {
        assertSiteExists(siteId);
        List<Record2<BuildingRecord, Integer>> rows = dsl
            .select(BUILDING, count(FLOOR.ID))
            .from(BUILDING)
            .leftJoin(FLOOR).on(FLOOR.BUILDINGID.eq(BUILDING.ID))
            .where(BUILDING.SITEID.eq(siteId))
            .groupBy(BUILDING.fields())
            .orderBy(BUILDING.SORTORDER.asc(), BUILDING.NAME.asc())
            .fetch();
        return rows.stream()
            .map(r -> BuildingDto.from(r.value1(), r.value2().longValue()))
            .toList();
    }

    public BuildingRecord getBuildingRecordById(String id) {
        BuildingRecord building = dsl.selectFrom(BUILDING).where(BUILDING.ID.eq(id)).fetchOne();
        if (building == null) {
            throw ApiException.notFound("Building not found");
        }
        return building;
    }

    public record CreateBuildingRequest(String name, String addressLine1, String addressLine2, String city,
                                         String state, String postalCode, String country, Double latitude,
                                         Double longitude, String status, String buildingType, String buildingCode,
                                         String description, Integer sortOrder) {
    }

    public BuildingRecord createBuilding(String siteId, CreateBuildingRequest req) {
        assertSiteExists(siteId);
        if (isBlank(req.name()) || isBlank(req.addressLine1()) || isBlank(req.city())
            || isBlank(req.state()) || isBlank(req.postalCode()) || isBlank(req.country())) {
            throw ApiException.badRequest("name, addressLine1, city, state, postalCode, and country are required");
        }
        BuildingRecord record = dsl.newRecord(BUILDING);
        record.setId(UUID.randomUUID().toString());
        record.setSiteid(siteId);
        record.setName(req.name().trim());
        record.setAddressline1(req.addressLine1().trim());
        record.setAddressline2(nullableTrim(req.addressLine2()));
        record.setCity(req.city().trim());
        record.setState(req.state().trim());
        record.setPostalcode(req.postalCode().trim());
        record.setCountry(req.country().trim());
        record.setLatitude(req.latitude());
        record.setLongitude(req.longitude());
        if (req.status() != null) record.setStatus(Entitystatus.valueOf(req.status().toUpperCase()));
        record.setBuildingtype(blankToNull(req.buildingType()));
        record.setBuildingcode(blankToNull(req.buildingCode()));
        record.setDescription(blankToNull(req.description()));
        record.setSortorder(req.sortOrder() != null ? req.sortOrder() : 0);
        record.setCreatedat(LocalDateTime.now());
        record.setUpdatedat(LocalDateTime.now());
        record.insert();
        record.refresh(); // pick up DB-side defaults (e.g. status) jOOQ doesn't return-and-populate automatically
        return record;
    }

    public record UpdateBuildingRequest(String name, String addressLine1, String addressLine2, String city,
                                         String state, String postalCode, String country, Double latitude,
                                         Double longitude, String status, String buildingType, String buildingCode,
                                         String description, Integer sortOrder) {
    }

    public BuildingRecord updateBuilding(String id, UpdateBuildingRequest req) {
        BuildingRecord record = getBuildingRecordById(id);
        boolean changed = false;
        if (req.name() != null) { record.setName(req.name().trim()); changed = true; }
        if (req.addressLine1() != null) { record.setAddressline1(req.addressLine1().trim()); changed = true; }
        if (req.addressLine2() != null) { record.setAddressline2(nullableTrim(req.addressLine2())); changed = true; }
        if (req.city() != null) { record.setCity(req.city().trim()); changed = true; }
        if (req.state() != null) { record.setState(req.state().trim()); changed = true; }
        if (req.postalCode() != null) { record.setPostalcode(req.postalCode().trim()); changed = true; }
        if (req.country() != null) { record.setCountry(req.country().trim()); changed = true; }
        if (req.latitude() != null) { record.setLatitude(req.latitude()); changed = true; }
        if (req.longitude() != null) { record.setLongitude(req.longitude()); changed = true; }
        if (req.buildingType() != null) { record.setBuildingtype(blankToNull(req.buildingType())); changed = true; }
        if (req.buildingCode() != null) { record.setBuildingcode(blankToNull(req.buildingCode())); changed = true; }
        if (req.description() != null) { record.setDescription(blankToNull(req.description())); changed = true; }
        if (req.sortOrder() != null) { record.setSortorder(req.sortOrder()); changed = true; }
        if (req.status() != null) { record.setStatus(Entitystatus.valueOf(req.status().toUpperCase())); changed = true; }
        if (!changed) {
            throw ApiException.badRequest("No fields to update");
        }
        record.setUpdatedat(LocalDateTime.now());
        record.update();
        return record;
    }

    public BuildingRecord deleteBuilding(String id) {
        BuildingRecord record = getBuildingRecordById(id);
        dsl.deleteFrom(BUILDING).where(BUILDING.ID.eq(id)).execute();
        return record;
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }

    private static String nullableTrim(String s) {
        return s != null ? s.trim() : null;
    }

    private static String blankToNull(String s) {
        if (s == null) return null;
        String trimmed = s.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}
